### ISSUE TITLE
CMake cannot identify CXX compiler if vcpkg has been integrated with Visual Studio.

### ISSUE THREAD

**Issue by rokoDev** [microsoft/vcpkg#15623]
**Describe the bug**
If `C:\Tools\vcpkg\vcpkg.exe integrate install` already has been called and then we try to generate Visual Studio solution by using CMake command: `cmake -G "Visual Studio 16 2019" -A x64 -T ClangCL ..`
CMake says that CXX compiler identification is unknown

**Environment**
- OS: Windows - 10.0.17763 - AMD64
- Compiler: Clang 10.0.0 with MSVC-like command-line

**To Reproduce**
1. `C:\Tools\vcpkg\vcpkg.exe integrate install`
2. `cd C:\projects\bugrc`
3. `mkdir build`
4. `cd build`
5. `cmake -G "Visual Studio 16 2019" -A x64 -T ClangCL ..`

This is [simple project](https://github.com/rokoDev/BugRC.git) which reproduces this bug on [Appveyor ](https://ci.appveyor.com/project/rokoDev/bugrc/builds/37240606/job/e45hqjxjk8ac79xj).

Actually Appveyor make call to `C:\Tools\vcpkg\vcpkg.exe integrate install` by default so to reproduce this error you do not need to call it explicitly. But if you wish to get rid of this error you need to uncomment and push **line 30**( `# - C:\Tools\vcpkg\vcpkg.exe integrate remove` ) in **appveyor.yml** which resides in the project root.

**Expected behavior**
I expect that in order to successfully generate Visual Studio solution from CMake only `cmake -G "Visual Studio 16 2019" -A x64 -T ClangCL ..` should be run without calling `C:\Tools\vcpkg\vcpkg.exe integrate remove` before that.

**Failure logs** 
**C:\projects\bugrc\build>cmake -G "Visual Studio 16 2019" -A x64 -T ClangCL ..**
`-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:7 (project):
  No CMAKE_CXX_COMPILER could be found.

-- Configuring incomplete, errors occurred!
See also "C:/projects/bugrc/build/CMakeFiles/CMakeOutput.log".
See also "C:/projects/bugrc/build/CMakeFiles/CMakeError.log".`

**This is "C:/projects/bugrc/build/CMakeFiles/CMakeOutput.log" content:**
`The system is: Windows - 10.0.17763 - AMD64`

**This is "C:/projects/bugrc/build/CMakeFiles/CMakeError.log" content:**
`Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" failed.
Compiler:  
Build flags: 
Id flags:  

The output was:
1
Microsoft (R) Build Engine version 16.8.2+25e4d540b for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/13/2021 10:38:42 AM.
Project "C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" on node 1 (default targets).
PrepareForBuild:
  Creating directory "Debug\".
  Creating directory "Debug\CompilerIdCXX.tlog\".
InitializeBuildStatus:
  Creating "Debug\CompilerIdCXX.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
VcpkgTripletSelection:
  Using triplet "x64-windows" from "C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\"
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\clang-cl.exe /c /I"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\include" /nologo /W0 /WX- /diagnostics:column /Od /D _MBCS /EHsc /MDd /GS /fp:precise /Fo"Debug\\" /Gd /TP -m64  CMakeCXXCompilerId.cpp
Link:
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\lld-link.exe /OUT:".\CompilerIdCXX.exe" /INCREMENTAL:NO /LIBPATH:"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib" /LIBPATH:"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\manual-link" kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib "C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib" /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG:FULL /PDB:".\CompilerIdCXX.pdb" /SUBSYSTEM:CONSOLE /DYNAMICBASE /NXCOMPAT /IMPLIB:".\CompilerIdCXX.lib"  Debug\CMakeCXXCompilerId.obj
lld-link : error : could not open 'C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib': no such file or directory [C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj]
Done Building Project "C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" (default targets) -- FAILED.

Build FAILED.

"C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" (default target) (1) ->
(Link target) -> 
  lld-link : error : could not open 'C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib': no such file or directory [C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.70

Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" failed.
Compiler:  
Build flags: 
Id flags:  

The output was:
1
Microsoft (R) Build Engine version 16.8.2+25e4d540b for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/13/2021 10:38:45 AM.
Project "C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" on node 1 (default targets).
PrepareForBuild:
  Creating directory "Debug\".
  Creating directory "Debug\CompilerIdCXX.tlog\".
InitializeBuildStatus:
  Creating "Debug\CompilerIdCXX.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
VcpkgTripletSelection:
  Using triplet "x64-windows" from "C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\"
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\clang-cl.exe /c /I"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\include" /nologo /W0 /WX- /diagnostics:column /Od /D _MBCS /EHsc /MDd /GS /fp:precise /Fo"Debug\\" /Gd /TP -m64  CMakeCXXCompilerId.cpp
Link:
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\lld-link.exe /OUT:".\CompilerIdCXX.exe" /INCREMENTAL:NO /LIBPATH:"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib" /LIBPATH:"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\manual-link" kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib "C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib" /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG:FULL /PDB:".\CompilerIdCXX.pdb" /SUBSYSTEM:CONSOLE /DYNAMICBASE /NXCOMPAT /IMPLIB:".\CompilerIdCXX.lib"  Debug\CMakeCXXCompilerId.obj
lld-link : error : could not open 'C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib': no such file or directory [C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj]
Done Building Project "C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" (default targets) -- FAILED.

Build FAILED.

"C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" (default target) (1) ->
(Link target) -> 
  lld-link : error : could not open 'C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib': no such file or directory [C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.71
`

Sorry for trouble if this is intended behaviour.

**Comment by ALittleDruid** [comment]
Add ninja.exe to your PATH environment

Run 'x64 Native Tools Command Prompt for VS 2019'
and then 

clang
cmake.exe -S .. -B . -GNinja "-DCMAKE_BUILD_TYPE:String=Debug" "-DCMAKE_CXX_COMPILER:STRING=C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/Llvm/x64/bin/clang++.exe" "-DCMAKE_C_COMPILER:STRING=C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/Llvm/x64/bin/clang.exe" "-DCMAKE_TOOLCHAIN_FILE:INTERNAL=D:/vcpkg/scripts/buildsystems/vcpkg.cmake"

msvc
cmake.exe -S .. -B . -GNinja "-DCMAKE_BUILD_TYPE:String=Debug" "-DCMAKE_CXX_COMPILER:STRING=C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29333/bin/HostX64/x64/cl.exe" "-DCMAKE_C_COMPILER:STRING=C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29333/bin/HostX64/x64/cl.exe" "-DCMAKE_TOOLCHAIN_FILE:INTERNAL=D:/vcpkg/scripts/buildsystems/vcpkg.cmake"

cmake.exe --build . --target all

**Comment by rokoDev** [comment]
@ALittleDruid your suggestion works but it does not allow to use advantages of Visual Studio IDE while compiling with clang. I tried to substitute **-GNinja** from **clang** example to **-G"Visual Studio 16 2019"** but this also switch current compiler from clang to msvc.

**Comment by JackBoosY** [comment]
Hmm... @dan-shaw can you please take a look?

Thanks.

**Comment by BillyONeal** [comment]
> C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\x64\bin\lld-link.exe /OUT:".\CompilerIdCXX.exe" /INCREMENTAL:NO /LIBPATH:"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib" /LIBPATH:"C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\manual-link" kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib "C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib" /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG:FULL /PDB:".\CompilerIdCXX.pdb" /SUBSYSTEM:CONSOLE /DYNAMICBASE /NXCOMPAT /IMPLIB:".\CompilerIdCXX.lib"  Debug\CMakeCXXCompilerId.obj
lld-link : error : could not open 'C:\Tools\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\*.lib': no such file or directory [C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj]
Done Building Project "C:\projects\bugrc\build\CMakeFiles\3.19.1\CompilerIdCXX\CompilerIdCXX.vcxproj" (default targets) -- FAILED.

It looks like `lld-link` isn't correctly replicating the behavior of `link.exe`, making this an LLVM bug. We nontheless may be able to work around it.

**Comment by RIscRIpt** [comment]
FYI original issue: https://github.com/Microsoft/vcpkg/issues/4359
A workaround was added in: https://github.com/microsoft/vcpkg/pull/4573

**TL;DR**: `vcpkg.props` uses `UseLldLink` variable to disable `VcpkgAutoLink`, however `UseLldLink` is defined in `Microsoft.Cpp.ClangCl.Common.props`, which is imported after `vcpkg.props`.

Reported upstream: *TODO*

### Details

1. In the latest VS 16.9.4 `vcpkg.props` has the following import tree:

```
This part may be wrong, but I keep it, in case you want to check it:
- Visual Studio project file
- Microsoft.Cpp.props
- Microsoft.Cpp.ToolsetLocation.props (imported in line 39 by Microsoft.Cpp.props)
- Platforms\x64\Platform.props (imported in line 50 by Microsoft.Cpp.ToolsetLocation.props)
- Microsoft.Cpp.Platform.props (imported in line 17 by Platform.props)
- Platforms\x64\PlatformToolsets\ClangCL\Toolset.props (imported in line 17 by Microsoft.Cpp.Platform.props)

The problem starts here:
- Microsoft.Cpp.MSVC.Toolset.x64.props (imported in line 16 by Toolset.props) !!! NOTICE THE LINE NUMBER
- Microsoft.Cpp.MSVC.Toolset.Common.props (imported in line 14 by Microsoft.Cpp.MSVC.Toolset.x64.props)
- Microsoft.Cpp.Common.props (imported in line 28 by Microsoft.Cpp.MSVC.Toolset.Common.props)
- $(LOCALAPPDATA)\vcpkg\vcpkg.user.props (imported in line 265 by Microsoft.Cpp.Common.props)
- scripts\buildsystems\msbuild\vcpkg.props
```

2. `UseLldLink` is defined by VS in `Microsoft.Cpp.ClangCl.Common.props`, which has the following import tree:

```
- Visual Studio project file
- Microsoft.Cpp.props
- Microsoft.Cpp.ToolsetLocation.props (imported in line 39 by Microsoft.Cpp.props)
- Platforms\x64\Platform.props (imported in line 50 by Microsoft.Cpp.ToolsetLocation.props)
- Microsoft.Cpp.Platform.props (imported in line 17 by Platform.props)
- Platforms\x64\PlatformToolsets\ClangCL\Toolset.props (imported in line 17 by Microsoft.Cpp.Platform.props)

The problem is here:
- Microsoft.Cpp.ClangCl.Common.props (imported in line 17 by Toolset.props) !!! NOTICE THE LINE NUMBER
```

Swapping line 16 with 17 seems to fix the problem, **however**, I'm not aware if `Microsoft.Cpp.ClangCl.Common.props` has depended variables from `Microsoft.Cpp.ClangCl.Common.props`.

**Comment by JackBoosY** [comment]
We should rename `UseLldLink` to avoid the name conflict.

**Comment by TENX-S** [comment]
Any updates on this ? Linking libs manually is terrible !

**Comment by RIscRIpt** [comment]
> Any updates on this ? Linking libs manually is terrible !

IIRC the problem is on the side of Visual Studio props files, not vcpkg.

**Comment by kgfly** [comment]
This is still repro with VS2022

**Comment by tonideleo** [comment]
I can confirm this is still happening. To make it work:
1) I have to manually specify the VS Clang Compiler to the cmake project.
2) Remove llvm from vcpkg to force VS use the integrated one
